File: addons/im_livechat/models/mail_channel_member.py
# Part of Odoo. See LICENSE file for full copyright and licensing details.

from datetime import datetime, timedelta

from odoo import api, models


class ChannelMember(models.Model):
    _inherit = 'mail.channel.member'

    @api.autovacuum
    def _gc_unpin_livechat_sessions(self):
        """ Unpin read livechat sessions with no activity for at least one day to
            clean the operator's interface """
        members = self.env['mail.channel.member'].search([
            ('is_pinned', '=', True),
            ('last_seen_dt', '<=', datetime.now() - timedelta(days=1)),
            ('channel_id.channel_type', '=', 'livechat'),
        ])
        sessions_to_be_unpinned = members.filtered(lambda m: m.message_unread_counter == 0)
        sessions_to_be_unpinned.write({'is_pinned': False})
        self.env['bus.bus']._sendmany([(member.partner_id, 'mail.channel/unpin', {'id': member.channel_id.id}) for member in sessions_to_be_unpinned])

    def _get_partner_data(self, fields=None):
        if self.channel_id.channel_type == 'livechat':
            data = {
                'active': self.partner_id.active,
                'id': self.partner_id.id,
                'is_public': self.partner_id.is_public,
            }
            if self.partner_id.user_livechat_username:
                data['user_livechat_username'] = self.partner_id.user_livechat_username
            else:
                data['name'] = self.partner_id.name
            if not self.partner_id.is_public:
                data['country'] = {
                    'code': self.partner_id.country_id.code,
                    'id': self.partner_id.country_id.id,
                    'name': self.partner_id.country_id.name,
                } if self.partner_id.country_id else [('clear',)]
            return data
        return super()._get_partner_data(fields=fields)

File: addons/im_livechat/static/tests/helpers/mock_server/models/mail_channel_member.js
/** @odoo-module **/

import '@mail/../tests/helpers/mock_server/models/mail_channel_member'; // ensure mail overrides are applied first

import { patch } from "@web/core/utils/patch";
import { MockServer } from "@web/../tests/helpers/mock_server";

patch(MockServer.prototype, 'im_livechat/models/mail_channel_member', {
    /**
     * @override
     */
    _mockMailChannelMember_GetPartnerData(ids) {
        const [member] = this.getRecords('mail.channel.member', [['id', 'in', ids]]);
        const [channel] = this.getRecords('mail.channel', [['id', '=', member.channel_id]]);
        const [partner] = this.getRecords('res.partner', [['id', '=', member.partner_id]], { active_test: false });
        if (channel.channel_type === 'livechat') {
            const data = {
                'id': partner.id,
                'is_public': partner.is_public,
            };
            if (partner.user_livechat_username) {
                data['user_livechat_username'] = partner.user_livechat_username;
            } else {
                data['name'] = partner.name;
            }
            if (!partner.is_public) {
                const [country] = this.getRecords('res.country', [['id', '=', partner.country_id]]);
                data['country'] = country
                    ? {
                        'code': country.code,
                        'id': country.id,
                        'name': country.name,
                    }
                    : [['clear']];
            }
            return data;
        }
        return this._super(ids);
    },
});
